File: Code/Java_Code/Chapter06_Collision/SphereCollision.java
import javax.swing.*;
import java.awt.*;
import javax.swing.border.BevelBorder;
import java.awt.event.*;
import javax.swing.Timer;

public class SphereCollision extends JFrame implements ActionListener
{
  private JTextField vx1TextField;
  private JTextField mass1TextField;
  private JTextField vx2TextField;
  private JTextField mass2TextField;
  private JTextField eTextField;

  private JLabel vx1Label;
  private JLabel mass1Label;
  private JLabel vx2Label;
  private JLabel mass2Label;
  private JLabel eLabel;

  private JButton startButton;
  private JButton resetButton;
  private JPanel drawingPanel;
  private GridBagConstraints gbc;

  //  These fields store sphere data and the coefficient 
  //  of restitution.
  private double vx1;
  private double x1;
  private double mass1;
  private double vx2;
  private double x2;
  private double mass2;
  private double e;            //  coefficient of restitution
  private double sphereRadius; //  radius of spheres

  //  These elements are used to control the execution
  //  speed of the game. Without them, the game would
  //  run too quickly.
  private GameUpdater gameUpdater;
  private Timer gameTimer;

  public SphereCollision() {

    //  Initialize sphere location, velocity, and
    //  coefficient of restitution.
    vx1 = 10.0;
    x1 = 5.0;
    mass1 = 10.0;
    vx2 = -5.0;
    x2 = 15.0;
    mass2 = 5.0;
    e = 0.9;
    sphereRadius = 2.0;

    //  Create a Timer object that will be used
    //  to slow the action down and an ActionListener
    //  that the Timer will call. The timeDelay variable
    //  is the time delay in milliseconds.
    gameUpdater = new GameUpdater();
    int timeDelay = 50;
    gameTimer = new Timer(timeDelay, gameUpdater);

    //  Create JTextField objects to input the velocity and
    //  mass of the sphere and the coefficient of restitution.
    vx1TextField = new JTextField("10.0",6);
    mass1TextField = new JTextField("10.0",6);
    vx2TextField = new JTextField("-5.0",6);
    mass2TextField = new JTextField("5.0",6);
    eTextField = new JTextField("0.9",6);

    //  Create some JLabels
    vx1Label = new JLabel("Sphere 1 velocity, m/s");
    mass1Label = new JLabel("Sphere 1 mass, kg");
    vx2Label = new JLabel("Sphere 2 velocity, m/s");
    mass2Label = new JLabel("Sphere 2 mass, kg");
    eLabel = new JLabel("Coefficient of restitution");

    //  Create a JButton that will start the balls moving
    startButton = new JButton("Start");
    startButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    startButton.setPreferredSize(new Dimension(60,35));
    startButton.addActionListener(this);

    //  Create a JButton that will update the drawing area.
    resetButton = new JButton("Reset");
    resetButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    resetButton.setPreferredSize(new Dimension(60,35));
    resetButton.addActionListener( new ActionListener() {
      public void actionPerformed(ActionEvent ae) {
        //  stop the timer.
        gameTimer.stop();

        //  Reset the sphere location, velocity, and
        //  coefficient of restitution.
        vx1 = 10.0;
        x1 = 5.0;
        mass1 = 10.0;
        vx2 = -5.0;
        x2 = 15.0;
        mass2 = 5.0;
        e = 0.9;

        //  Update the display.
        updateDisplay();
      }  
    });

    //  Create a JTextArea that will display the results
    drawingPanel = new JPanel();
    drawingPanel.setPreferredSize(new Dimension(301, 201));

    //  Place components on a panel using a GridBagLayout
    JPanel westPanel = new JPanel();
    GridBagLayout gridBagLayout1 = new GridBagLayout();
    westPanel.setLayout(gridBagLayout1);

    int col;
    int row;
    int numCol = 1;
    int numRow = 1;
    Insets insets = new Insets(5, 3, 5, 3);

    row = 0;
    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vx1Label, gbc);

    col = 2;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vx1TextField, gbc);

    row = 1;
    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(mass1Label, gbc);

    col = 2;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(mass1TextField, gbc);

    row = 2;
    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vx2Label, gbc);

    col = 2;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vx2TextField, gbc);

    row = 3;
    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(mass2Label, gbc);

    col = 2;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(mass2TextField, gbc);

    row = 4;
    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(eLabel, gbc);

    col = 2;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(eTextField, gbc);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(startButton, gbc);

    row = 1;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(resetButton, gbc);

    westPanel.add(vx1Label);
    westPanel.add(vx1TextField);
    westPanel.add(mass1Label);
    westPanel.add(mass1TextField);
    westPanel.add(vx2Label);
    westPanel.add(vx2TextField);
    westPanel.add(mass2Label);
    westPanel.add(mass2TextField);
    westPanel.add(eLabel);
    westPanel.add(eTextField);
    westPanel.add(startButton);
    westPanel.add(resetButton);

    //  The drawing panel.
    JPanel southPanel = new JPanel();
    GridBagLayout gridBagLayout2 = new GridBagLayout();
    southPanel.setLayout(gridBagLayout2);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.CENTER,
                 GridBagConstraints.NONE, 
                 new Insets(10, 10, 10, 20), 0, 0);
    gridBagLayout2.setConstraints(drawingPanel, gbc);

    southPanel.add(drawingPanel);

    //  Add the JPanel objects to the content pane
    getContentPane().setLayout(new BorderLayout());
    getContentPane().add(westPanel, BorderLayout.WEST);
    getContentPane().add(southPanel, BorderLayout.SOUTH);

    //  Add a title to the JFrame, size it, and make it visible.
    setTitle("Linear Collision Simulator");
    setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    setBounds(100,100,500,500);
    setVisible(true);

    //  Update the GUI display
    updateDisplay(); 
  }

  //  The actionPerformed() method is called when 
  //  the "Start" button is pressed. 
  public void actionPerformed(ActionEvent event) {
    
    //  Get the initial quantities from the textfields.
    vx1 = Double.parseDouble(vx1TextField.getText());
    mass1 = Double.parseDouble(mass1TextField.getText());
    vx2 = Double.parseDouble(vx2TextField.getText());
    mass2 = Double.parseDouble(mass2TextField.getText());
    e = Double.parseDouble(eTextField.getText());

    //  Update the display
    updateDisplay();

    //  Start the box sliding using a Timer object
    //  to slow down the action.
    gameTimer.start();
  }

  //  This method redraws the GUI display.
  private void updateDisplay() {
    Graphics g = drawingPanel.getGraphics();
    int width = drawingPanel.getWidth() - 1;
    int height = drawingPanel.getHeight() - 1;

    g.clearRect(0, 0, width, height);
    g.setColor(Color.WHITE);
    g.fillRect(0, 0, width, height); 

    //  Update the position of the spheres on the screen.
    g.setColor(Color.BLACK);
    int xPosition = (int)(10.0*(x1 - sphereRadius));
    g.drawOval(xPosition, 80, 2*(int)(10.0*sphereRadius), 
               2*(int)(10.0*sphereRadius));

    xPosition = (int)(10.0*(x2 - sphereRadius));
    g.drawOval(xPosition, 80, 2*(int)(10.0*sphereRadius), 
               2*(int)(10.0*sphereRadius));

  }

  public static void main(String args[]) {
    SphereCollision gui = new SphereCollision();
  }

  //  This ActionListener is called by the Timer
  class GameUpdater implements ActionListener {
    public void actionPerformed(ActionEvent event) {

      //  Determine if a collision occurs and if it
      //  does, change the velocties of the spheres.
      //  A collision occurs if the distance between the
      //  centers of the spheres is less than twice their
      //  radius
      double distance = Math.abs(x1 - x2);
      double tmp = 1.0/(mass1 + mass2);

      if ( distance <= 2.0*sphereRadius ) {
        double newVx1 = (mass1 - e*mass2)*vx1*tmp + 
                        (1.0 + e)*mass2*vx2*tmp;
        double newVx2 = (1.0 + e)*mass1*vx1*tmp +
                        (mass2 - e*mass1)*vx2*tmp;
        vx1 = newVx1;
        vx2 = newVx2;
      }

      //  Compute the new location of the spheres. 
      double timeIncrement = 0.07;
      x1 = x1 + timeIncrement*vx1;
      x2 = x2 + timeIncrement*vx2;

      //  Update the display
      updateDisplay();

      //  If either of the spheres hits the outer edge
      //  of the display, stop the simulation.
      Graphics g = drawingPanel.getGraphics();
      int width = drawingPanel.getWidth() - 1;
      int height = drawingPanel.getHeight() - 1;

      if ( 10.0*(x1 + sphereRadius) > width || 
           10.0*(x1 - sphereRadius) < 0.0 ||
           10.0*(x2 + sphereRadius) > width || 
           10.0*(x2 - sphereRadius) < 0.0 ) {
        gameTimer.stop();
      }
    }
  }
}

File: Code/Csharp_Code/Chapter06_Collision/SphereCollision.cs
using System;
using System.Windows.Forms;
using System.Drawing;

public class SphereCollision : Form
{
  private Label vx1Label;
  private Label mass1Label;
  private Label vx2Label;
  private Label mass2Label;
  private Label eLabel;

  private TextBox vx1TextBox;
  private TextBox mass1TextBox;
  private TextBox vx2TextBox;
  private TextBox mass2TextBox;
  private TextBox eTextBox;

  private Button startButton;
  private Button resetButton;

  private Panel drawingPanel;

  //  These fields store sphere data and the coefficient 
  //  of restitution.
  private double vx1;
  private double x1;
  private double mass1;
  private double vx2;
  private double x2;
  private double mass2;
  private double e;            //  coefficient of restitution
  private double sphereRadius; //  radius of spheres

  //  The Timer is used to control the execution speed
  //  of the game.
  private Timer gameTimer;

  public SphereCollision() {

    //  Initialize sphere location, velocity, and
    //  coefficient of restitution.
    vx1 = 10.0;
    x1 = 5.0;
    mass1 = 10.0;
    vx2 = -5.0;
    x2 = 15.0;
    mass2 = 5.0;
    e = 0.9;
    sphereRadius = 2.0;

    //  Create a Timer object that will be used
    //  to slow the action down.
    gameTimer = new Timer();
    gameTimer.Interval = 50;  //  delay in milliseconds.
    gameTimer.Tick += new EventHandler(ActionPerformed);

    //  Create some Labels
    vx1Label = new Label();
    vx1Label.Text = "Sphere 1 velocity, m/s";
    vx1Label.Font = new Font(vx1Label.Font, FontStyle.Bold);
    vx1Label.Top = 30;
    vx1Label.Left = 10;
    vx1Label.Width = 130;

    mass1Label = new Label();
    mass1Label.Text = "Sphere 1 mass, kg";
    mass1Label.Font = new Font(mass1Label.Font, FontStyle.Bold);
    mass1Label.Top = 60;
    mass1Label.Left = 10;
    mass1Label.Width = 130;

    vx2Label = new Label();
    vx2Label.Text = "Sphere 2 velocity, m/s";
    vx2Label.Font = new Font(vx2Label.Font, FontStyle.Bold);
    vx2Label.Top = 90;
    vx2Label.Left = 10;
    vx2Label.Width = 130;

    mass2Label = new Label();
    mass2Label.Text = "Sphere 2 mass, kg";
    mass2Label.Font = new Font(mass2Label.Font, FontStyle.Bold);
    mass2Label.Top = 120;
    mass2Label.Left = 10;
    mass2Label.Width = 130;

    eLabel = new Label();
    eLabel.Text = "Coefficient of restitution";
    eLabel.Font = new Font(eLabel.Font, FontStyle.Bold);
    eLabel.Top = 150;
    eLabel.Left = 10;
    eLabel.Width = 130;

    //  Create TextBox objects to display the outcome.
    vx1TextBox = new TextBox();
    vx1TextBox.Width = 50;
    vx1TextBox.Text = "10.0";
    vx1TextBox.AutoSize = true;
    vx1TextBox.Top = vx1Label.Top;
    vx1TextBox.Left = 150;

    mass1TextBox = new TextBox();
    mass1TextBox.Width = 50;
    mass1TextBox.Text = "10.0";
    mass1TextBox.AutoSize = true;
    mass1TextBox.Top = mass1Label.Top;
    mass1TextBox.Left = 150;

    vx2TextBox = new TextBox();
    vx2TextBox.Width = 50;
    vx2TextBox.Text = "-5.0";
    vx2TextBox.AutoSize = true;
    vx2TextBox.Top = vx2Label.Top;
    vx2TextBox.Left = 150;

    mass2TextBox = new TextBox();
    mass2TextBox.Width = 50;
    mass2TextBox.Text = "5.0";
    mass2TextBox.AutoSize = true;
    mass2TextBox.Top = mass2Label.Top;
    mass2TextBox.Left = 150;

    eTextBox = new TextBox();
    eTextBox.Width = 50;
    eTextBox.Text = "0.9";
    eTextBox.AutoSize = true;
    eTextBox.Top = eLabel.Top;
    eTextBox.Left = 150;

    //  Create Button objects 
    int buttonHeight = 30;
    int buttonWidth = 50;
    int buttonLeft = 240;

    startButton = new Button();
    startButton.Text = "Start";
    startButton.Height = buttonHeight;
    startButton.Width = buttonWidth;
    startButton.Top = 50;
    startButton.Left = buttonLeft;
    startButton.Click += new EventHandler(StartButtonClicked);

    resetButton = new Button();
    resetButton.Text = "Reset";
    resetButton.Height = buttonHeight;
    resetButton.Width = buttonWidth;
    resetButton.Top = 110;
    resetButton.Left = buttonLeft;
    resetButton.Click += new EventHandler(ResetButtonClicked);

    //  Create a drawing panel.
    drawingPanel = new Panel();
    drawingPanel.Width = 301;
    drawingPanel.Height = 201;
    drawingPanel.Left = 70;
    drawingPanel.Top = 220;
    drawingPanel.BorderStyle = BorderStyle.FixedSingle;

    //  Add the GUI components to the Form
    this.Controls.Add(vx1Label);
    this.Controls.Add(vx2Label);
    this.Controls.Add(mass1Label);
    this.Controls.Add(mass2Label);
    this.Controls.Add(eLabel);
    this.Controls.Add(vx1TextBox);
    this.Controls.Add(mass1TextBox);
    this.Controls.Add(vx2TextBox);
    this.Controls.Add(mass2TextBox);
    this.Controls.Add(eTextBox);
    this.Controls.Add(startButton);
    this.Controls.Add(resetButton);
    this.Controls.Add(drawingPanel);

    // Set the size and title of the form
    this.Width = 500;
    this.Height = 500;
    this.Text = "Sphere Collision";

    //  Center the form on the screen and make
    //  it visible.
    this.StartPosition = FormStartPosition.CenterScreen;
    this.Visible = true;

    //  Update the GUI display
    UpdateDisplay();
  }

  //  Event handling method for the "Start" button
  public void StartButtonClicked(object source, EventArgs evt) {

    //  Extract initial data from the textfields.
    vx1 = Convert.ToDouble(vx1TextBox.Text);
    mass1 = Convert.ToDouble(mass1TextBox.Text);
    vx2 = Convert.ToDouble(vx2TextBox.Text);
    mass2 = Convert.ToDouble(mass2TextBox.Text);
    e = Convert.ToDouble(eTextBox.Text);

    //  Start the box sliding using a Timer object
    //  to slow down the action.
    gameTimer.Start();
  }

  //  Event handling method for the "Reset" button
  public void ResetButtonClicked(object source, EventArgs evt) {
    //  stop the timer.
    gameTimer.Stop();

    //  Reset the sphere location, velocity, and
    //  coefficient of restitution.
    vx1 = 10.0;
    x1 = 5.0;
    mass1 = 10.0;
    vx2 = -5.0;
    x2 = 15.0;
    mass2 = 5.0;
    e = 0.9;

    //  Update the display.
    UpdateDisplay();
  }

  //  This method redraws the GUI display.
  private void UpdateDisplay() {
    Graphics g = drawingPanel.CreateGraphics();
    int width = drawingPanel.Width - 1;
    int height = drawingPanel.Height - 1;

    //  Clear the current display.
    g.Clear(drawingPanel.BackColor);

    //  Update the position of the spheres on the screen.
    Pen blackPen = new Pen(Color.Black, 2);
    int xPosition = (int)(10.0*(x1 - sphereRadius));
    g.DrawEllipse(blackPen, xPosition, 80, 2*(int)(10.0*sphereRadius), 
               2*(int)(10.0*sphereRadius));

    xPosition = (int)(10.0*(x2 - sphereRadius));
    g.DrawEllipse(blackPen, xPosition, 80, 2*(int)(10.0*sphereRadius), 
               2*(int)(10.0*sphereRadius));

    //  Clean up the Graphics object.
    g.Dispose();
  }

  //  This method is called by the Timer every 0.05 seconds.
  public void ActionPerformed(object source, EventArgs evt) {
    //  Determine if a collision occurs and if it
    //  does, change the velocties of the spheres.
    //  A collision occurs if the distance between the
    //  centers of the spheres is less than twice their
    //  radius
    double distance = Math.Abs(x1 - x2);
    double tmp = 1.0/(mass1 + mass2);

    if ( distance <= 2.0*sphereRadius ) {
      double newVx1 = (mass1 - e*mass2)*vx1*tmp + 
                      (1.0 + e)*mass2*vx2*tmp;
      double newVx2 = (1.0 + e)*mass1*vx1*tmp +
                      (mass2 - e*mass1)*vx2*tmp;
      vx1 = newVx1;
      vx2 = newVx2;
    }

    //  Compute the new location of the spheres. 
    double timeIncrement = 0.07;
    x1 = x1 + timeIncrement*vx1;
    x2 = x2 + timeIncrement*vx2;

    //  Update the display
    UpdateDisplay();

    //  If either of the spheres hits the outer edge
    //  of the display, stop the simulation.
    int width = drawingPanel.Width - 1;

    if ( 10.0*(x1 + sphereRadius) > width || 
         10.0*(x1 - sphereRadius) < 0.0 ||
         10.0*(x2 + sphereRadius) > width || 
         10.0*(x2 - sphereRadius) < 0.0 ) {
      gameTimer.Stop();
    }
  }

  static void Main() {
    Application.Run(new SphereCollision());
  }
}
